allennlp/modules/seq2seq_encoders/pytorch_seq2seq_wrapper.py

```diff
@@ -29,76 +29,93 @@ class PytorchSeq2SeqWrapper(Seq2SeqEncoder):
-        if num_valid < batch_size:
-            mask = mask.clone()
-            mask[:, 0] = 1
-
-        sequence_lengths = get_lengths_from_binary_sequence_mask(mask)
-        sorted_inputs, sorted_sequence_lengths, restoration_indices = sort_batch_by_length(inputs,
-                                                                                           sequence_lengths)
-        packed_sequence_input = pack_padded_sequence(sorted_inputs,
-                                                     sorted_sequence_lengths.data.tolist(),
-                                                     batch_first=True)
-
-        # Actually call the module on the sorted PackedSequence.
-        packed_sequence_output, _ = self._module(packed_sequence_input, hidden_state)
+        packed_sequence_output, final_states, restoration_indices, num_valid = \
+            self.sort_and_run_forward(self._module, inputs, mask, hidden_state)
+
         unpacked_sequence_tensor, _ = pad_packed_sequence(packed_sequence_output, batch_first=True)
 
-        # We sorted by length, so if there are invalid rows that need to be zeroed out
-        # they will be at the end.
+        # Some RNNs (GRUs) only return one state as a Tensor.  Others (LSTMs) return two.
+        # If one state, use a single element list to handle in a consistent manner below.
+        if not isinstance(final_states, (list, tuple)) and self._stateful:
+            final_states = [final_states]
+
+        # Add back invalid rows.
         if num_valid < batch_size:
-            unpacked_sequence_tensor[num_valid:, :, :] = 0.
+            _, length, output_dim = unpacked_sequence_tensor.size()
+            zeros = unpacked_sequence_tensor.data.new(batch_size - num_valid, length, output_dim).fill_(0)
+            unpacked_sequence_tensor = torch.cat([unpacked_sequence_tensor, zeros], 0)
+
+            # The states also need to have invalid rows added back.
+            if self._stateful:
+                new_states = []
+                for state in final_states:
+                    num_layers, _, state_dim = state.size()
+                    zeros = state.data.new(num_layers, batch_size - num_valid, state_dim).fill_(0)
+                    new_states.append(torch.cat([state, zeros], 1))
+                final_states = new_states
 
         # It's possible to need to pass sequences which are padded to longer than the
         # max length of the sequence to a Seq2SeqEncoder. However, packing and unpacking
         # the sequences mean that the returned tensor won't include these dimensions, because
         # the RNN did not need to process them. We add them back on in the form of zeros here.
         sequence_length_difference = total_sequence_length - unpacked_sequence_tensor.size(1)
         if sequence_length_difference > 0:
-            zeros = unpacked_sequence_tensor.data.new(batch_size, sequence_length_difference,
+            zeros = unpacked_sequence_tensor.data.new(batch_size,
```

Comment:
It's the same `add_rows` operation here, just on a different dimension.  And you're making `zeros` a `Variable` here, but not above - I assume `torch.cat` returns a `Variable` if any of its inputs is a `Variable`?